src/kde/kde.cc

```diff
@@ -30,124 +30,437 @@ namespace rtt_kde {
+          }
+        }
+      }
     }
+  }
+
+  // normalize the integrated weight contributions
+  for (size_t i = 0; i < local_size; i++) {
+    Check(normal[i] > 0.0);
+    result[i] /= normal[i];
+  }
+
+  return result;
+}
 
-    rtt_c4::global_sum(global_x_position.data(), size);
-    rtt_c4::global_sum(global_distribution.data(), size);
-    rtt_c4::global_sum(global_conservation);
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief KDE reconstruction done in logarithmic data space
+ * 
+ * \pre The local reconstruction data is passed into this function which
+ * includes the original data distribution, its spatial position, and the
+ * optimal bandwidth to be used at each point. The original data distribution
+ * is transformed into log space prior and post reconstruction. This is helpful
+ * for strongly peaked data and should be exact for exponential distributions.
+ *
+ * \param[in] distribution original data to be reconstructed
+ * \param[in] one_over_bandwidth inverse bandwidth size to be used at each data location
+ * \param[in] qindex quick_index class to be used for data access.
+ * \param[in] discontinuity_cutoff maximum size of value discrepancies to include in the reconstruction
+ * \return final local KDE function distribution reconstruction
+ *
+ * \post the local reconstruction of the original data is returned.
+ */
+std::vector<double>
+kde::log_reconstruction(const std::vector<double> &distribution,
+                        const std::vector<std::array<double, 3>> &one_over_bandwidth,
+                        const quick_index &qindex, const double discontinuity_cutoff) const {
+  size_t dim = qindex.dim;
+  Require(dim < 3 && dim > 0);
+  const size_t local_size = distribution.size();
+  Require(qindex.locations.size() == local_size);
+  Require(one_over_bandwidth.size() == local_size);
 
+  // used for the zero accumulation conservation
+  std::vector<double> result(local_size, 0.0);
+  std::vector<double> normal(local_size, 0.0);
+  double min_value = *std::min_element(distribution.begin(), distribution.end());
+  double max_value = *std::max_element(distribution.begin(), distribution.end());
+  double log_bias = fabs(min_value) + (max_value - min_value);
+  if (qindex.domain_decomposed) {
+
+    rtt_c4::global_min(min_value);
+    rtt_c4::global_min(max_value);
+
+    std::vector<double> ghost_distribution(qindex.local_ghost_buffer_size);
+    qindex.collect_ghost_data(distribution, ghost_distribution);
+    std::vector<std::array<double, 3>> ghost_one_over_bandwidth(qindex.local_ghost_buffer_size,
+                                                                {0.0, 0.0, 0.0});
+    qindex.collect_ghost_data(one_over_bandwidth, ghost_one_over_bandwidth);
+
+    log_bias = fabs(min_value) + (max_value - min_value);
+    // if the log bias is zero the answer must be zero everywhere
+    if (!(log_bias > 0.0))
+      return result;
```

Comment:
Is it possible to create a test case that exercises this early return?

Reply:
Yap. I'll add a zero data reconstruction to catch these parts of the code.